include/exadg/operators/interior_penalty_parameter.h

```diff
@@ -44,48 +44,87 @@ calculate_penalty_parameter(AlignedVector<VectorizedArray<Number>> & array_penal
-  FEValues<dim> fe_values(mapping, fe, quadrature, update_JxW_values);
+  // Go through all faces and compute their area. Since FEFaceEvaluation
+  // visits faces exactly once, we need to accumulate results into a global
+  // vector where we can call compress and update ghost values
+  unsigned int const                         level = matrix_free.get_mg_level();
+  LinearAlgebra::distributed::Vector<Number> surface_areas;
+  if(auto * tria = dynamic_cast<const parallel::TriangulationBase<dim> *>(
+       &matrix_free.get_dof_handler(dof_index).get_triangulation()))
+  {
+    if(level == numbers::invalid_unsigned_int)
+      surface_areas.reinit(tria->global_active_cell_index_partitioner().lock());
+    else
+      surface_areas.reinit(tria->global_level_cell_index_partitioner(level).lock());
+  }
+  else
+  {
+    surface_areas.reinit(
+      matrix_free.get_dof_handler(dof_index).get_triangulation().n_active_cells());
+  }
 
-  QGauss<dim - 1>   face_quadrature(degree + 1);
-  FEFaceValues<dim> fe_face_values(mapping, fe, face_quadrature, update_JxW_values);
+  // Lambda function to simplify access to cell index for the two cases
+  auto const get_cell_index = [level](typename Triangulation<dim>::cell_iterator const & cell) {
+    if(level == numbers::invalid_unsigned_int)
+      return cell->global_active_cell_index();
+    else
+      return cell->global_level_cell_index();
+  };
 
-  for(unsigned int i = 0; i < n_cells; ++i)
   {
-    for(unsigned int v = 0; v < matrix_free.n_active_entries_per_cell_batch(i); ++v)
-    {
-      typename DoFHandler<dim>::cell_iterator cell = matrix_free.get_cell_iterator(i, v, dof_index);
-      fe_values.reinit(cell);
+    FEFaceEvaluation<dim, -1, 0, 1, Number> fe_eval_m(matrix_free, true, dof_index);
 
-      // calculate cell volume
-      Number volume = 0;
-      for(unsigned int q = 0; q < quadrature.size(); ++q)
-      {
-        volume += fe_values.JxW(q);
-      }
+    unsigned int const n_face_batches =
+      matrix_free.n_inner_face_batches() + matrix_free.n_boundary_face_batches();
+    for(unsigned int face = 0; face < n_face_batches; ++face)
+    {
+      fe_eval_m.reinit(face);
+      VectorizedArray<Number> face_area = 0.0;
+      for(unsigned int q = 0; q < fe_eval_m.n_q_points; ++q)
+        face_area += fe_eval_m.JxW(q);
 
-      // calculate surface area
-      Number surface_area = 0;
-      for(const unsigned int f : cell->face_indices())
+      for(unsigned int v = 0; v < matrix_free.n_active_entries_per_face_batch(face); ++v)
       {
-        fe_face_values.reinit(cell, f);
-        Number const factor = (cell->at_boundary(f) && !cell->has_periodic_neighbor(f)) ? 1. : 0.5;
-        for(unsigned int q = 0; q < face_quadrature.size(); ++q)
+        // inner face: adjust weight and also write to the "exterior" face
+        if(face < matrix_free.n_inner_face_batches())
         {
-          surface_area += fe_face_values.JxW(q) * factor;
+          face_area[v] *= 0.5;
```

Comment:
Why don't you separate this code from the code below that adds to `surface_areas` (i.e. shift the `if()` clause to line 102 and briefly explain why `*=0.5`)? The main motivation for this is that multiplication does not have to be done separately for each `v`, so I think this change would improve readability a lot. This would also very naturally resolve the issue that I raised above regarding the comment (that was difficult to understand for me).